### ISSUE TITLE
site-package content get installed into build_env instead of host_env when cross-compiling

### ISSUE THREAD

**Issue by Tobias-Fischer** [conda-forge/cross-python-feedstock#24]
Original issue: https://github.com/mamba-org/boa/issues/158

Within the RoboStack project, we observed that cross-compiling (host: osx, target: osx-arm64) a recipe failed. We were able to pin down the issue to cross-python, in particular those lines in the activation script: https://github.com/conda-forge/cross-python-feedstock/blob/41083676420cef1f476a0bde41632cdf7334279b/recipe/activate-cross-python.sh#L33-L34
While the original issue in https://github.com/mamba-org/boa/issues/158 was observed when using boa and osx cross-compile, I was since able to replicate the same problem with `conda build` and linux->linux-aarch64 cross-compilation.

Basically, the `site-packages` directory in the host_env ends up being empty, and its contents are instead copied into the build_env. However, in the case of vtk, the cmake script refers to a library in `$HOST/lib/python3.8/site-packages`:

```
CMake Error at /home/tobias/anaconda3/conda-bld/pcl_1621892989199/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/cmake/vtk-9.0/VTKPython-targets.cmake:395 (message):
  The imported target "VTK::vtkCommonCorePython" references the file

     "/home/tobias/anaconda3/conda-bld/pcl_1621892989199/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/python3.7/site-packages/vtkmodules/vtkCommonCore.so"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/home/tobias/anaconda3/conda-bld/pcl_1621892989199/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/cmake/vtk-9.0/VTKPython-targets.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /home/tobias/anaconda3/conda-bld/pcl_1621892989199/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/lib/cmake/vtk-9.0/vtk-config.cmake:147 (include)
  cmake/pcl_find_vtk.cmake:25 (find_package)
  CMakeLists.txt:377 (include)
```

vtk does not show up with `conda list` after activating the host_env, but it does show up in the build_env (however note that channel is pypi rather than conda):
```
# packages in environment at /Users/tobias/opt/anaconda3/conda-bld/ros_1621565257168/_build_env:
...
vtk                       9.0.1                    pypi_0    pypi
...
```

@wolfv's comment in the original issue:
> I think these lines from the cross-python activation script might be relevant:
> 
> https://github.com/conda-forge/cross-python-feedstock/blob/41083676420cef1f476a0bde41632cdf7334279b/recipe/activate-cross-python.sh#L33-L34
> 
> Maybe @isuruf has an idea (sorry to bother you!). I think for us the problem is that we need to link to the shared libs in site-packages.
> 
> However we don't try to execute these python executables so there won't be any issues with executing the wrong arch... maybe we could improve this activation script in a way to not remove some `dylib` or `so` files depending on some configured env var ...
> 
> @Tobias-Fischer we should probably move this issue to the cross-python feedstock

**Comment by wolfv** [comment]
It's not quite clear to me why the `.so / .dylib` files have to be removed from the host env (since they are properly compiled for the host arch etc.).

Maybe one way to fix this on our side would be to replace all `$PREFIX/lib/python3.x/site-packages` references in the CMake files with `$BUILD_PREFIX/lib/python...`.

Or we could have an env var that we can set to prevent this activation script from removing the vtk .so files?

**Comment by isuruf** [comment]
> It's not quite clear to me why the .so / .dylib files have to be removed from the host env (since they are properly compiled for the host arch etc.).

Because python will try to load from `$PREFIX` first and that would fail. If we let python to load from `$BUILD_PREFIX`, then some paths might be wrong.
For eg: if we let python load numpy from ${PREFIX}, it will fail and if we let python load numpy from ${BUILD_PREFIX}, then the blas paths will be for the BUILD_PREFIX and building scipy would fail.

Solution was to use files from $PREFIX except for the python extensions which comes from BUILD_PREFIX.
If we had used PREFIX/lib/pythonx.y/site-packages to be the place to have this, then PREFIX might get new files from BUILD_PREFIX and get packaged. Safest option was to do this in BUILD_PREFIX.

**Comment by Tobias-Fischer** [comment]
@wolfv:
So I was able to cross-build PCL (linux -> linux-aarch64; conda-build) by adding this to the build.sh:

```
PYTHON_VERSION_WITH_DOT="${CONDA_PY:0:1}.${CONDA_PY:1}"
find $PREFIX/lib/cmake -type f -exec sed -i "s~\${_IMPORT_PREFIX}/lib/python$PYTHON_VERSION_WITH_DOT/site-packages~$BUILD_PREFIX/lib/python$PYTHON_VERSION_WITH_DOT/site-packages~g" {} +
```

However, the pcl-conversions recipe (osx -> osx-arm64; boa) still fails:
```
if [[ "$CONDA_BUILD_CROSS_COMPILATION" == "1" ]]; then
  find $PREFIX/lib/cmake -type f -exec sed -i "s~\${_IMPORT_PREFIX}/lib/python$ROS_PYTHON_VERSION/site-packages~$BUILD_PREFIX/lib/python$ROS_PYTHON_VERSION/site-packages~g" {} +
fi
```

In this pcl-conversions case, the so files don't get copied into the build_env, whereas in the pcl case they do. Any ideas?

**Comment by Tobias-Fischer** [comment]
Hi @wolfv @isuruf - do you have any more ideas re this issue?

**Comment by matthewwardrop** [comment]
I'm also having issues with cross-compiling from linux-64 -> linux-aarch64 due to the deletion of the files from $PREFIX/lib/python<version>/site-packages; which causes a later running of the `setup.py` script to fail due to `setuptools` being missing (among other things). What is the suggested remedy for this?

**Comment by isuruf** [comment]
@matthewwardrop, which feedstock?

**Comment by matthewwardrop** [comment]
@isuruf Apologies, I should have provided a bit more context. The package in question is an internal proprietary package, but will be open sourced in the fullness of time. I was attempting to use this package to build internal conda packages for distribution internally until we are ready to open source things.

Assuming that this is a supported use-case, please see below for more details.

Our build looks like:

meta.yaml
```
{% set data = load_setup_py_data() %}
{% set base_name = "my_package" %}
{% set version = data['version'] | replace('-rc.', 'rc') %}

package:
  name: "{{ name }}"
  version: "{{ version }}"

about:
  home: internal_url
  license: PROPRIETARY
  license_family: PROPRIETARY
  summary: |
    ...

source:
  path: ..

outputs:
  - name: "{{ base_name }}"
    version: "{{ version | replace('-rc.', 'rc') }}"
    requirements:
      build:
        - {{ compiler('cxx') }}
        - python {{ python }}.*                  # [build_platform != target_platform]
        - cross-python_{{ target_platform }}     # [build_platform != target_platform]
        - pybind11 >=2.6                         # [build_platform != target_platform]
        - pybind11-abi                           # [build_platform != target_platform]
      host:
        - python {{ python }}.*
        - pip
        - setupmeta >=2.7.12

        # Required for arrow pointer tool compilation
        - pybind11 >=2.6
        - arrow-cpp {{ arrow }}.*
        - pyarrow {{ arrow }}.*
      run:
        - docstring_parser >=0.6
        - interface_meta
        - numpy
        - pandas >=1.2.0,!=1.3.0  # Indexing regression in 1.3.0 is fixed in 1.3.1
        - pyarrow {{ arrow }}.*
        - python {{ python }}.*
        - rpy2 >=3.4.0,<3.5
        - scipy
        - spec-classes >=1.1.0
        - statsmodels
        - sympy
        - tzlocal
    build:
      number: 0
      script: python -m pip install --no-deps --ignore-installed -vv .
```

conda_build_config
```
python:
  - 3.9
arrow:
  - 3.0.0
numpy:
  - 1.18
_openmp_mutex:
  - 4.5=0_gnu
pin_run_as_build:
  arrow_cpp: x.x
  pyarrow: x.x
  _openmp_mutex: x.x.x.x
ignore_version:
  - numpy
target_platform:
  # - linux-64   # [linux]
  - linux-aarch64  # [linux]
  # - osx-64     # [osx]
  # - osx-arm64  # [osx]
```

The build itself succeeds with this configuration on Linux-64 for aarch64, but then bails out in a later step due to the site-packages directory having removed setuptools (which I now note is needed to properly parse the meta.yaml due to _load_setup_py_data). I will do some testing tomorrow to see if hard-coding the version fixes things, though it is nice not to have to do that.

**Comment by isuruf** [comment]
Can't help you further without a reproducer. Sorry.

**Comment by ndevenish** [comment]
I've also just hit this issue. I'm working with a dependency that includes header files, and the location of the header files is in the `site-packages/` folder, and there is a separately installed "environment" file that lists the header locations.

Because the site-packages folder has been moved to the build environment: https://github.com/conda-forge/cross-python-feedstock/blob/b01685b23ab70ab6d445d80336d709df130f3616/recipe/activate-cross-python.sh#L57-L59

 the build failed. To solve this, I've just added a special case into my `build.sh`:
```
if [[ $CONDA_TOOLCHAIN_BUILD != $CONDA_TOOLCHAIN_HOST ]]; then
    CFLAGS="-isystem $BUILD_PREFIX/lib/python$PY_VER/site-packages $CXXFLAGS"
    CXXFLAGS="-isystem $BUILD_PREFIX/lib/python$PY_VER/site-packages $CXXFLAGS"
fi
```

I'm mainly just documenting this - not really asking for a solution; while it'd be ideal if I didn't have to do anything special in the `build.sh` script for cross-compiling, realistically it's always going to be a special case; it just took me a while to work out what the cross- build was doing and why.
